File: frontend/lib/contracts/sales.ts
/**
 * Sales & Billing Contracts
 * Defines types for sales transactions, invoices, and cart items
 */

export interface CartItem {
  productId: string;
  name: string;
  price: number;
  quantity: number;
  stock: number;
}

export interface CartItemWithTotal extends CartItem {
  itemTotal: number;
}

export type PaymentMethod = "cash" | "credit";

export interface Invoice {
  id: string;
  shopName: string;
  shopContact?: string;
  invoiceId: string;
  dateTime: string;
  items: CartItemWithTotal[];
  subtotal: number;
  discount: number;
  total: number;
  paymentMethod: PaymentMethod;
  customerId?: string;
  customerName?: string;
  dueAmount?: number;
  itemCount: number;
}

export interface SalesHistory {
  id: string;
  invoiceId: string;
  total: number;
  paymentMethod: PaymentMethod;
  timestamp: string;
  itemCount: number;
  customerName?: string;
}

export interface CreateSaleRequest {
  items: CartItem[];
  discount: number;
  paymentMethod: PaymentMethod;
  customerId?: string;
  customerName?: string;
  dueAmount?: number;
}

export interface SalesHistoryResponse {
  sales: SalesHistory[];
}

export interface CreateSaleResponse {
  invoice: Invoice;
  success: boolean;
}

export interface Customer {
  id: string;
  name: string;
  email?: string;
  phone?: string;
  totalCredit: number;
  dueAmount: number;
  createdAt: string;
  updatedAt: string;
}

export interface CreateCustomerRequest {
  name: string;
  email?: string;
  phone?: string;
}

File: backend/app/schemas/sales.py
from pydantic import BaseModel, Field


class CartItem(BaseModel):
    productId: str
    name: str
    price: float = Field(ge=0)
    quantity: int = Field(ge=1)
    stock: int = Field(ge=0)


class CartItemWithTotal(CartItem):
    itemTotal: float = Field(ge=0)


class PaymentMethod(str):
    """Payment method: cash or credit"""
    pass


class Invoice(BaseModel):
    id: str
    shopName: str
    shopContact: str | None = None
    invoiceId: str
    dateTime: str
    items: list[CartItemWithTotal]
    subtotal: float = Field(ge=0)
    discount: float = Field(ge=0)
    total: float
    paymentMethod: str  # "cash" or "credit"
    customerId: str | None = None
    customerName: str | None = None
    dueAmount: float | None = None
    creditUntil: str | None = None
    itemCount: int = Field(ge=0)


class SalesHistory(BaseModel):
    id: str
    invoiceId: str
    total: float
    paymentMethod: str  # "cash" or "credit"
    timestamp: str
    itemCount: int = Field(ge=0)
    customerName: str | None = None


class CreateSaleRequest(BaseModel):
    items: list[CartItem] = Field(min_items=1)
    discount: float = Field(ge=0)
    paymentMethod: str  # "cash" or "credit"
    customerId: str | None = None
    customerName: str | None = None
    dueAmount: float | None = None
    creditUntil: str | None = None


class SalesHistoryResponse(BaseModel):
    sales: list[SalesHistory]


class CreateSaleResponse(BaseModel):
    invoice: Invoice
    success: bool
